prow/tide/tide.go

```diff
@@ -0,0 +1,289 @@
+	return &Controller{
+		log: l,
+		ghc: ghc,
+		kc:  kc,
+		ca:  ca,
+	}
+}
+
+// Sync runs one sync iteration.
+func (c *Controller) Sync() error {
+	ctx := context.Background()
+	c.log.Info("Building tide pool.")
+	var pool []pullRequest
+	for _, q := range c.ca.Config().Tide.Queries {
+		prs, err := c.search(ctx, q)
+		if err != nil {
+			return err
+		}
+		pool = append(pool, prs...)
+	}
+	var pjs []kube.ProwJob
+	var err error
+	if len(pool) > 0 {
+		c.log.Info("Listing ProwJobs.")
+		pjs, err = c.kc.ListProwJobs(nil)
+		if err != nil {
+			return err
+		}
+	}
+	sps, err := c.dividePool(pool, pjs)
+	if err != nil {
+		return err
+	}
+	for _, sp := range sps {
+		if err := c.syncSubpool(ctx, sp); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+type simpleState string
+
+const (
+	noneState    simpleState = "none"
+	pendingState simpleState = "pending"
+	successState simpleState = "success"
+)
+
+func toSimpleState(s kube.ProwJobState) simpleState {
+	if s == kube.TriggeredState || s == kube.PendingState {
+		return pendingState
+	} else if s == kube.SuccessState {
+		return successState
+	}
+	return noneState
+}
+
+func pickSmallestNumber(prs []pullRequest) pullRequest {
+	smallestNumber := -1
+	var smallestPR pullRequest
+	for _, pr := range prs {
+		if smallestNumber < 0 || int(pr.Number) < smallestNumber {
+			smallestNumber = int(pr.Number)
+			smallestPR = pr
+		}
+	}
+	return smallestPR
+}
+
+// TODO(spxtr): Batch merges. These complicate the logic further. One
+// particularly nasty complication is that we want to merge PRs even if the
+// serial test for a PR failed but the batch succeeded. This means that the PR
+// will need to be in the tide pool despite a failing status.
+func (c *Controller) syncSubpool(ctx context.Context, sp subpool) error {
+	c.log.Infof("%s/%s %s: %d PRs, %d PJs.", sp.org, sp.repo, sp.branch, len(sp.prs), len(sp.pjs))
+	var presubmits []string
+	for _, ps := range c.ca.Config().Presubmits[sp.org+"/"+sp.repo] {
+		if ps.SkipReport || !ps.AlwaysRun || !ps.RunsAgainstBranch(sp.branch) {
+			continue
+		}
+		presubmits = append(presubmits, ps.Name)
+	}
+	var successes []pullRequest
+	var pendings []pullRequest
+	var nones []pullRequest
+	for _, pr := range sp.prs {
+		// Accumulate the best result for each job.
+		psStates := make(map[string]simpleState)
+		for _, pj := range sp.pjs {
+			if pj.Spec.Refs.Pulls[0].Number != int(pr.Number) {
+				continue
+			}
+			name := pj.Spec.Job
+			oldState := psStates[name]
+			newState := toSimpleState(pj.Status.State)
+			if oldState == noneState || oldState == "" {
+				psStates[name] = newState
+			} else if oldState == pendingState && newState == successState {
+				psStates[name] = successState
+			}
+		}
+		// The overall result is the worst of the best.
+		overallState := successState
+		for _, ps := range presubmits {
+			if s, ok := psStates[ps]; s == noneState || !ok {
+				overallState = noneState
```

Comment:
nit: `break` here, and you will not need `overallState == successState` check below :-)